### ISSUE TITLE
ITestRunner.Load() fails loading Nunit30Settings.xml when called in parallel

### ISSUE THREAD

**Issue by bretehlert** [nunit/nunit#1270]
When using nunit.engine.api 3.0.1 and executing test discovery in parallel in different processes on the same machine we get the following error loading Nunit30Settings.xml:

NUnit.Engine.NUnitEngineException: Error loading settings C:\Users\s_datservice\AppData\Local\NUnit\Nunit30Settings.xml. The process cannot access the file 'C:\Users\s_datservice\AppData\Local\NUnit\Nunit30Settings.xml' because it is being used by another process. ---> System.IO.IOException: The process cannot access the file 'C:\Users\s_datservice\AppData\Local\NUnit\Nunit30Settings.xml' because it is being used by another process.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize)
   at System.Xml.XmlDownloadManager.GetStream(Uri uri, ICredentials credentials, IWebProxy proxy, RequestCachePolicy cachePolicy)
   at System.Xml.XmlUrlResolver.GetEntity(Uri absoluteUri, String role, Type ofObjectToReturn)
   at System.Xml.XmlTextReaderImpl.OpenUrlDelegate(Object xmlResolver)
   at System.Runtime.CompilerServices.RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup(TryCode code, CleanupCode backoutCode, Object userData)
   at System.Threading.CompressedStack.Run(CompressedStack compressedStack, ContextCallback callback, Object state)
   at System.Xml.XmlTextReaderImpl.OpenUrl()
   at System.Xml.XmlTextReaderImpl.Read()
   at System.Xml.XmlLoader.Load(XmlDocument doc, XmlReader reader, Boolean preserveWhitespace)
   at System.Xml.XmlDocument.Load(XmlReader reader)
   at System.Xml.XmlDocument.Load(String filename)
   at NUnit.Engine.Internal.SettingsStore.LoadSettings()
   --- End of inner exception stack trace ---
   at NUnit.Engine.Internal.SettingsStore.LoadSettings()
   at NUnit.Engine.Services.SettingsService.StartService()
   at NUnit.Engine.Services.ServiceManager.StartServices()
   at NUnit.Agent.NUnitTestAgent.Main(String[] args)
System.Exception: Unable to acquire remote process agent
   at NUnit.Engine.Runners.ProcessRunner.LoadPackage()
   at NUnit.Engine.Runners.AbstractTestRunner.Load()
   at NUnit.Engine.Runners.MasterTestRunner.LoadPackage()
   at NUnit.Engine.Runners.AbstractTestRunner.EnsurePackageIsLoaded()
   at NUnit.Engine.Runners.MasterTestRunner.NUnit.Engine.ITestRunner.Explore(TestFilter filter)

My calling code looks like this:
			using (var engine = TestEngineActivator.CreateInstance(privateCopy: true))
			using (var runner = engine.GetRunner(new TestPackage(path)))
			{
				var xml = runner.Explore(TestFilter.Empty);
			}

It appears that this code writes to the Nunit30Settings.xml file every time it runs, thus creating the access error when one process is writing to the settings file and another is loading it.

The contents that are saved to the file are as follows:
<?xml version="1.0"?>
<NUnitSettings>
  <Settings />
</NUnitSettings>

**Comment by CharliePoole** [comment]
Are you creating multiple engines in your code?

**Comment by CharliePoole** [comment]
The instance of our SettingsService that is created in the agents is readonly. It should only be possible to write the settings file from the main runner process in which the engine runs. If you were to create multiple instances of the engine in that process, each one would be able to write to the file.

**Comment by bretehlert** [comment]
I have found that the TestEngine instance creates a writeable SettingsService and therefore the settings are saved each time we dispose the engine.

We are running multiple processes in parallel, each process creates one engine using the nunit.engine.api.

**Comment by CharliePoole** [comment]
That's not how NUnit itself works or how the engine is designed to be used. There should be one primary instance of the engine, which controls all other instances. The TestActivator is designed to create that instance. Other intances are created by ProcessRunner launching nunit-agent.exe.

What are trying to accomplish? Perhaps I can make some suggestions. Frankly, it's hard to imagine what might cause you to need your own executable for the agent processes, but if there is some functionality missing, we may be able to add it.

**Comment by bretehlert** [comment]
If you say that NUnit does not support creating more than one engine at a time on one computer, we can adapt our implementation to that constraint. I suggest you document that clearly on the Nunit API page https://github.com/nunit/dev/wiki/Test-Engine-API.

That said, I ask, why does it makes sense to you that each time I dispose an engine it saves an empty settings file to disk? Can we have an option to create an engine with read-only settings? Or only persist if settings changed? At the moment we have hacked a workaround to this issue by casting the TestEngine instance up, reflecting out the private _writeable field, and setting it to false. Now using multiple engines in parallel on the same system works fine.

I am implementing an adaptor for NUnit 3 for our own continuous integration system that runs over 50 gated check-ins per day on six different branches of a product with over 500,000 unit tests in 1000 assemblies (100 cpu hours to execute one run). So, we tend to focus on scalability, parallelizing as much as we can, on each node and across many nodes, and using process isolation for long running stability and cross-version support.

**Comment by CharliePoole** [comment]
I'll look at clarifying the docs, but actually that's not what I'm saying at all. It's the SettingsService that should only have one writable instance at a time. When you initialize an Engine, you give it a bunch of services, usually including a SettingsService. Its the SettingsService that can be writable or not.

Again, it's the SettingsService that, if it's writable, flushes its content to disk when the service is stopped. If it's not writable, it naturally doesn't do so. If you initialize your engine with a non-writable SettingsService in the first place, you won't have the problem at all.

[SettingsService could, of course, check to see if anything is dirty before saving its content. But that would involve a bit of extra code that we decided to omit.]

What you are doing sounds like something that NUnit could do as designed, using multiple Runners provided by a single engine rather than multiple (primary) engines. (As opposed to the second-level engines created by the nunit-agent program.) Just create a test package and get an ITestRunner for each job as it comes along. Using default settings in the package will give you process isolation and automatically select the target framework and bitness of the process to be used.

However, if you are already invested in the approach you are taking, you should stop using the TestEngineActivator to create your engines. It is designed for "mass market" use and creates an engine with the services that are usually needed by almost all users. You should look at how the nunit-agent program creates and initializes the engine instead. It actually instantiates the TestEngine class and initializes it with a set of services that are needed in an agent runner.

**Comment by bretehlert** [comment]
Ok, thanks for all of the information. I hadn't thought about bypassing TestEngineActivator to create an engine with pre-initialized services. Doing this combined with changing the implementation to use one engine with many runners when appropriate should work fine.

**Comment by CharliePoole** [comment]
FYI, NUnit creates it's primary agent using TestEngineActivator. The rest are created "by hand" in the agent executable, which is little more than a communications stub tacked on to the engine itself.

The primary engine includes the TestAgencyService, which functions as a server and is connected to by the agents, its clients. The "service" it provides is telling the agent what tests to run. The agent then runs the tests and exits.

Hope this is useful to you.

**Comment by ammachado** [comment]
To whom it may concern, I'm having this exception while Visual Studio 2015 is discovering the tests to be run when I trigger the "Test" > "Run" > "All Tests" in the menu.

**Comment by bbi-yoyo** [comment]
I have the same problem as @ammachado.

**Comment by CharliePoole** [comment]
@ammachado @bbi-yoyo This issue is closed. It's about a problem that the OP was having when calling NUnit programmatically through our Test Engine API. If that's what either of you are doing, and if the answer I gave to the OP doesn't help you, we can reopen this.

OTOH, if you are having a problem simply running NUnit as a user, that's something else. Please create a new issue and give us a good description of what we have to do in order to see the problem.

**Comment by bbi-yoyo** [comment]
@CharliePoole it happened once while I was opening a solution. Closing and reopening the solution it's not happening any more, so I'm willing to chalk it up to gremlins for now.

**Comment by mwadams** [comment]
Did anyone open an issue for this? I get this problem all the time on VS2015 Update 2 (w/ latest patch) and NUnit Adapter 3.2.0.0

**Comment by CharliePoole** [comment]
@mwaddams No new issue has been opened related to having multiple processes trying to open the NUnit30Settings.xml file at least as far as I know. Is that what you are seeing wtih the NUnit 3 adapter? If so, you should open an issue there. This closed issue was not related to the adapter at all.

**Comment by a-jaeger** [comment]
I have the same problem with Adapter 3.2.0.0 and Visual Studio 2015 and open an [issue](https://github.com/nunit/nunit3-vs-adapter/issues/190).
